pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/NullableSingleInputAggregationFunction.java

```diff
@@ -0,0 +1,263 @@
+   * @param nullIndexIterator an int iterator that returns values in ascending order whose min value is 0.
+   *                          Rows are considered null if and only if their index is emitted.
+   * @param initialAcum the initial value of the accumulator
+   * @param <A> The type of the accumulator
+   */
+  public <A> A foldNotNull(int length, @Nullable IntIterator nullIndexIterator, A initialAcum, Reducer<A> reducer) {
+    A acum = initialAcum;
+    if (!_nullHandlingEnabled || nullIndexIterator == null || !nullIndexIterator.hasNext()) {
+      return reducer.apply(initialAcum, 0, length);
+    }
+
+    int prev = 0;
+    while (nullIndexIterator.hasNext() && prev < length) {
+      int nextNull = Math.min(nullIndexIterator.next(), length);
+      if (nextNull > prev) {
+        acum = reducer.apply(acum, prev, nextNull);
+      }
+      prev = nextNull + 1;
+    }
+    if (prev < length) {
+      acum = reducer.apply(acum, prev, length);
+    }
+    return acum;
+  }
+
+  /**
+   * Returns an iterator that returns the union of the null bitmaps of valSet1 and valSet2.
+   */
+  public IntIterator orNullIterator(BlockValSet valSet1, BlockValSet valSet2) {
+    if (!_nullHandlingEnabled) {
+      return EmptyIntIterator.INSTANCE;
+    } else {
+      RoaringBitmap nullBlock1 = valSet1.getNullBitmap();
+      RoaringBitmap nullBlock2 = valSet2.getNullBitmap();
+      if (nullBlock1 == null) {
+        return nullBlock2 == null ? EmptyIntIterator.INSTANCE : nullBlock2.getIntIterator();
+      } else if (nullBlock2 == null) {
+        return nullBlock1.getIntIterator();
+      } else {
+        return new MinIntIterator(nullBlock1.getIntIterator(), nullBlock2.getIntIterator());
+      }
+    }
+  }
+
+  public static class EmptyIntIterator implements IntIterator {
+
+    public static final EmptyIntIterator INSTANCE = new EmptyIntIterator();
+
+    private EmptyIntIterator() {
+    }
+
+    @Override
+    public IntIterator clone() {
+      return this;
+    }
+
+    @Override
+    public boolean hasNext() {
+      return false;
+    }
+
+    @Override
+    public int next() {
+      throw new NoSuchElementException();
+    }
+  }
+
+  /**
+   * An iterator that returns the minimum of the values returned by two other iterators.
+   *
+   * This means that at any point, the next value returned by this iterator is the minimum of the next values in the
+   * set defined by both sub-iterators.
+   */
+  public static class MinIntIterator implements IntIterator {
+    private final IntIterator _it1;
+    private final IntIterator _it2;
+    private int _next1 = -1;
+    private int _next2 = -1;
+
+    /**
+     * @param it1 it has to iterate in ascending order and the min value is 0
+     * @param it2 it has to iterate in ascending order and the min value is 0
+     */
+    public MinIntIterator(IntIterator it1, IntIterator it2) {
+      _it1 = it1;
+      _it2 = it2;
+    }
+
+    @Override
+    public IntIterator clone() {
```

Comment:
Do we need this override? Seems strange to me as we should never clone it

Reply:
IntIterator.close is abstract in RoaringBitmaps, so we have to implement it. Anyway, I've just remove this code to introduce it in the percentille PR, which is the one that actually uses it